### PR TITLE
[NOT NEEDED] [Traceable FSDP2] Improve FSDPParam._unsharded_param tracing

### PR DESCRIPTION
**Pull request by yf225**
Please see detailed changes in the code comment. This PR helps simplify the forward graph output by removing unnecessary tensors (e.g. `self.all_gather_outputs`) from it.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128927

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 1]
```diff
@@ -520,8 +527,22 @@ def free_unsharded_param(self) -> None:
         ):
             free_storage(tensor)
         if ca.compiled_autograd_enabled:
-            self.all_gather_outputs = []
-            self._unsharded_inner_tensors = []
+            # Under compile, `self.all_gather_outputs` and `self._unsharded_inner_tensors` are emptied
+            # right after every AllGather. So we need to free `self.unsharded_param` memory by
+            # looking into its own tensor storage instead.
+            unsharded_param_data = (
+                self.unsharded_param.data
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:534]: I think generally, calling `.data` is not specifically what we want. Why not call `.detach()`? Or actually, why do we even need to detach from the graph at all (is this needed for tracing)?

E.g., can we case on wrapper subclass before calling `.data`?

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 2]
```diff
@@ -520,8 +527,22 @@ def free_unsharded_param(self) -> None:
...
+            )  # Unpack nn.Parameter to get underlying tensor data
+            is_subclass = (
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:536]: nit: probably can reuse the `is_traceable_wrapper_subclass`?
https://github.com/pytorch/pytorch/blob/e6d4451ae8987bf8d6ad85eb7cde685fac746f6f/torch/utils/_python_dispatch.py#L268

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 1 above)
Comment (yf225) [reply torch/distributed/_composable/fsdp/_fsdp_param.py]: sg - `self.unsharded_param` is a `nn.Parameter` which is a subclass on its own, so I think we might need `.data` to get its underlying tensor (could still be a subclass (e.g. DTensor) or plain tensor). I changed it so that we only do `.data` access for the subclass case

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 2 above)
Comment (yf225) [reply torch/distributed/_composable/fsdp/_fsdp_param.py]: yes thanks - `is_traceable_wrapper_subclass` is much better
